### PR TITLE
Add more physics options to the Scene importer

### LINKED ISSUE #6932: Add "Physics Material Override" option for generated bodies in the Advanced Import Settings dialog
### Describe the project you are working on

AA Indie first person shooter.

Thousands of environment assets.

### Describe the problem or limitation you are having in your project

Currently the "Physics Material Override" property is not exposed when importing a mesh file.

Physics Body type, size and shape are all exposed and configurable so it makes sense to also expose the override.

Physics Material Override is used in-game not only to determine how the surface type responds but we use it to determine the surface type of a collider for footstep sounds, bullet hit particles, decals etc. Therefore every mesh in the game needs to be assigned the correct physics material. "Concrete, Metal, Water" etc.

This is an important feature for projects with many hundreds/thousands of assets since currently, the only way to make sure this field is populated correctly is to create a new inherited scene for every mesh in the game which adds up quickly, reducing productivity and polluting the filesystem.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

image

My proposal is to add a new option to the Scene Importer's Physics section that will allow you to specify the material for a given body at the import level.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

This should be relatively simple as the importer is already generating the physics body for us, it is just one additional propery we need to populate on the new Node.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

No.

### Is there a reason why this should be core and not an add-on in the asset library?

This modifies an existing core component to add a small additional feature. It has wide ranging use cases and I believe is an important improvements to the core.

### PR DESCRIPTION
**Pull request by EMBYRDEV**
Implements and closes https://github.com/godotengine/godot-proposals/issues/6932

image

### CODE REVIEW

**Comment by EMBYRDEV** [comment]
I should probably also add the option to set the collision layer and mask. These are important for the same reason exposing PMO is (not having to create hundreds of inherited scenes just to configure things properly)

**Comment by EMBYRDEV** [comment]
image

Done!

**Comment by EMBYRDEV** [comment]
After some additional discussion, I'm going to expose some additional options on meshes too to streamline the import process.
- Cast Shadow
- Lightmap Scale
- LOD Bias
- Visibility Ranges
- Layer

**Comment by fire** [comment]
I want to support this, but trying to get people to review.

**Comment by EMBYRDEV** [comment]
> I want to support this, but trying to get people to review.

It's not quite ready for review yet since speaking to Calinou I plan on adding a bunch of extra options.

That being said Calinou seems to be on board but I'm not sure if I'll need additional approval for this to be pulled in.

**Comment by fire** [comment]
Can you consider splitting the feature? It's easier to merge smaller changes.

**Comment by EMBYRDEV** [comment]
> Can you consider splitting the feature? It's easier to merge smaller changes.

Sure thing, In that case please feel free to get this change reviewed as I think the physics options belong in a single PR.

**Comment by AThousandShips** [comment]
We're in feature freeze now, and the milestone is 4.x, so I think this will have to wait for 4.2

**Comment by EMBYRDEV** [comment]
Is there anything else needing done before looking at getting this merged?

**Review by Calinou — APPROVED** [review]
Tested locally (rebased on top of `master` 851bc640ddf7225a4b52bb15cc50c249df119953), it works as expected.

**Testing project:** [test_pr_77533.zip](https://github.com/godotengine/godot/files/12065293/test_pr_77533.zip)

File: editor/import/resource_importer_scene.cpp
Comment (rburing) [inline editor/import/resource_importer_scene.cpp:1249]: This could be `if (pmo.is_valid())` (same applies to the other cases), but no biggie.

https://github.com/godotengine/godot/blob/202e4b2c1e7f8b25738b93d0e4d5066453d3edf3/core/object/ref_counted.h#L201-L202

**Review by rburing — APPROVED** [review]
Makes sense. Makes the advanced importer a lot more useful with just a tiny bit of code.

**Pull request merged**
